### PR TITLE
Recoverabiltiy use custom type for exception passing

### PR DESCRIPTION
**Pull request by tmasternak**
Moving await from Exception instance in ErrorContext. This makes it easier for transport seams to store exception information outside of process (e.g. ASB can store it in headers) and makes more explicit what data is required by recoverability. In addition this changes aligns us better with .Net Core which does not support exception serialization in 1.0 release.

Connects to https://github.com/Particular/V6Launch/issues/69

### CODE REVIEW

**Comment by tmasternak** [comment]
@Particular/nservicebus-maintainers please review. Thx!

File: src/NServiceBus.Core/Utils/ExceptionHeaderHelper.cs [hunk 1]
```diff
@@ -1,55 +1,47 @@
...
             }
 
-            headers["NServiceBus.ExceptionInfo.HelpLink"] = e.HelpLink;
-            headers["NServiceBus.ExceptionInfo.Message"] = e.GetMessage().Truncate(16384);
-            headers["NServiceBus.ExceptionInfo.Source"] = e.Source;
+            headers["NServiceBus.ExceptionInfo.HelpLink"] = exceptionInfo.HelpLink;
+            headers["NServiceBus.ExceptionInfo.Message"] = exceptionInfo.Message.Truncate(16384);
+            headers["NServiceBus.ExceptionInfo.Source"] = exceptionInfo.Source;
 
             if (legacyStacktrace)
```
Comment (tmasternak) [inline src/NServiceBus.Core/Utils/ExceptionHeaderHelper.cs]: @SimonCropp do you think we can remove support for legacy stack traces. See: https://github.com/Particular/NServiceBus/issues/2447

**Comment by tmasternak** [comment]
@tmasternak Let's remove the message truncation on the error headers. Now transport seams can manage that when creating ErrorInfo. I've update PoA to make sure we move the current truncation to RabbitMQ as agreed in https://github.com/Particular/NServiceBus/issues/2956

File: src/NServiceBus.Core/Utils/ExceptionHeaderHelper.cs (hunk 1 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/Utils/ExceptionHeaderHelper.cs]: @SimonCropp our proposal would be to deprecate it entirely. The app.config key is not documented.

Comment (SimonCropp) [inline src/NServiceBus.Core/Utils/ExceptionHeaderHelper.cs]: yep no complaints

File: src/NServiceBus.Core/Recoverability/RecoverabilityExecutor.cs [hunk 2]
```diff
@@ -63,11 +64,11 @@ Task<ErrorHandleResult> PerformRecoverabilityAction(ErrorContext errorContext)
         {
             var message = errorContext.Message;
 
-            Logger.Info($"First Level Retry is going to retry message '{message.MessageId}' because of an exception:", errorContext.Exception);
+            Logger.Info($"First Level Retry is going to retry message '{message.MessageId}' because of an exception:" + Environment.NewLine + errorContext.ExceptionInfo.StackTrace);
```
Comment (tmasternak) [inline src/NServiceBus.Core/Recoverability/RecoverabilityExecutor.cs]: Let's add string interpolation here and make sure we do not duplicate code

File: src/NServiceBus.Core/Recoverability/RecoverabilityExecutor.cs [hunk 3]
```diff
@@ -3,10 +3,10 @@
     using System;
     using System.Threading.Tasks;
     using Logging;
-    using Transport;
 
     class RecoverabilityExecutor
     {
+        //TODO: figure out proper logging for ExceptionInfo
```
Comment (tmasternak) [inline src/NServiceBus.Core/Recoverability/RecoverabilityExecutor.cs]: this is obsolete

**Comment by danielmarbach** [comment]
@tmasternak and I decided to close this PR without merging. The downside for the extension points to only have ExceptionInfo are far too high. Especially when considering that they'd use the ability to cast to a specific exception, the inner exception is not available etc.

**Pull request closed without merging**